dask_match/core.py

```diff
@@ -616,7 +625,245 @@ def optimize_expr(expr):
             expr = replace_all(expr, replacement_rules)
     finally:
         _defer_to_matchpy = False
+
+    if fuse:
+        # Apply fusion
+        expr = _blockwise_fusion(expr)
+
     return expr
 
 
 from dask_match.reductions import Count, Max, Min, Mode, Size, Sum
+
+
+## Utilites for Expr fusion
+
+
+def replace_nodes(expr, nodes_to_replace: dict):
+    """Replace specific nodes in an Expr tree
+
+    `nodes_to_replace` corresponds to a `dict` mapping
+    of old `Expr` objects to new `Expr` objects.
+    """
+    if not nodes_to_replace:
+        # Nothing to replace
+        return expr
+    elif expr in nodes_to_replace:
+        # Already at targetted expr
+        return nodes_to_replace[expr]
+
+    new_operands = []
+    new_count = 0
+    for operand in expr.operands:
+        if isinstance(operand, Expr) and operand in nodes_to_replace:
```

Comment:
I suggest that we support substituting non-exprs as well

Reply:
The existance of non-Expr objects in the operands is an anti-pattern since it means the language isn't closed. So I think it would rather make sense to avoid the case that non-Expr things can show up here, unless that is really crucial. It makes writing generic type-safe traversals and rewrites very hard.

Reply:
In general I agree with you, but in the specific case of lightweight languages embedded in other languages (like Python) I think that being impure ends up working better.   To make this concrete I'll ask the question:

*What would you recommend for operands like filenames and numbers?*

If you suggest that we make a `String` type and an `Int` type then I'll suggest that that is likely more heavyweight than what I personally would like to see.

Reply:
I think it makes sense to have a `Literal(value)` type (or if you will `Opaque(...)`) that doesn't have any operands. If doing things like replacement you still need to match on the object and dispatch based on the contents of `value`, but you no longer have to special-case it if you don't care.

IME, while it is tempting to avoid recreating types that already exist, this is a decision that typically becomes painful down the line, since every traversal and rewrite algorithm that is subsequently written has to remember that not every operand of every expr is an expr. This can then only be handled by isinstance checks which is fragile because you have an open untagged union of potential types (so it's easy to write `isinstance(expr, (str, int))` and then later for someone to shove a float in, or whatever). At least with the `Literal` object you control the surface area of this kind of checking.

That's not to say that one can't add sugar to the constructors such that literal python types are not wrapped up, but I think the internal representation should have an Expr object for them.

An alternative approach is to store python literals as "non-child" data, but that comes with drawbacks too if you want to rewrite there.